File: backend/src/main/java/br/com/calciolari/datahub/sales/api/SaleController.java
package br.com.calciolari.datahub.sales.api;

import java.time.LocalDateTime;
import java.util.UUID;

import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import br.com.calciolari.datahub.sales.api.SaleDtos.SaleDetail;
import br.com.calciolari.datahub.sales.api.SaleDtos.SaleSummary;
import br.com.calciolari.datahub.sales.application.SaleQueryService;
import br.com.calciolari.datahub.shared.api.PageParams;
import br.com.calciolari.datahub.shared.api.PageResponse;

@RestController
@RequestMapping("/api/sales")
public class SaleController {

	private final SaleQueryService saleQueryService;

	public SaleController(SaleQueryService saleQueryService) {
		this.saleQueryService = saleQueryService;
	}

	@GetMapping
	public PageResponse<SaleSummary> list(
			@RequestParam(required = false) UUID productId,
			@RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from,
			@RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime to,
			@RequestParam(required = false) Integer page,
			@RequestParam(required = false) Integer size) {
		return saleQueryService.list(productId, from, to, PageParams.page(page), PageParams.size(size));
	}

	@GetMapping("/{id}")
	public SaleDetail get(@PathVariable UUID id) {
		return saleQueryService.get(id);
	}
}

File: backend/src/Calciolari.DataHub/Sales/Api/SaleController.cs
using Calciolari.DataHub.Sales.Application;
using Calciolari.DataHub.Shared.Api;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Calciolari.DataHub.Sales.Api;

[ApiController]
[Route("api/sales")]
[Authorize(Roles = "VIEWER,IMPORTER,ADMIN")]
public sealed class SaleController : ControllerBase
{
    private readonly SaleQueryService _saleQueryService;

    public SaleController(SaleQueryService saleQueryService)
    {
        _saleQueryService = saleQueryService;
    }

    [HttpGet]
    public PageResponse<SaleSummary> List(
        [FromQuery] Guid? productId,
        [FromQuery] DateTime? from,
        [FromQuery] DateTime? to,
        [FromQuery] int? page,
        [FromQuery] int? size) =>
        _saleQueryService.List(productId, from, to, PageParams.Page(page), PageParams.Size(size));

    [HttpGet("{id:guid}")]
    public SaleDetail Get(Guid id) => _saleQueryService.Get(id);
}
